### ISSUE TITLE
Trapped exit messages 

### ISSUE THREAD

**Issue by lostbean** [purerl/purescript-erl-process#2]
I looking for a way to represent trapped exit messages:

`{'EXIT', Pid, normal} | {'EXIT', Pid, kill} | {'EXIT', Pid, Other}`

The ideal purescript representation would be, IMO, something like this:

```
data ExitMsg = EXIT Pid ExitReason
data ExitReason = Normal | Kill | Other String
``` 

One, apparently solvable, issue is conversion of the data constructor `EXIT` to the atom `eXIT`. We could have better heuristics for this conversion in the compiler. For instance, `EXIT -> 'EXIT'`, `Exit -> exit` and `ExitB -> exitB`.

Any idea how to implement this?

**Comment by nwolverson** [comment]
At the lowest level I'd represent the specific atoms via the FFI with an `foreign import` and use https://github.com/purerl/purescript-erl-tuples, rather than rely on the data constructor representation happening to align - and then add an `Atom` type at some point to avoid having to bind each atom via FFI.

Maybe keep your suggestion in mind if it becomes too tedious/keeps cropping up?

**Comment by lostbean** [comment]
It is still not clear for me:

1. How to create an sum type using `erl-tuples` that would have an runtime representation like this `{'EXIT', Pid, normal} | {'EXIT', Pid, kill} | {'EXIT', Pid, Other}`?
2. Is there any internal representation of atoms? Something like `Prim.Atom`?
3. By adding `Atom`, do you mean add an opaque data type `data Atom :: *` and a function `atom :: String -> Atom`?

Could you provide a simple example of how this low level representation would work?

**Comment by nwolverson** [comment]
```
foreign import data Atom :: *
foreign import atom :: String -> Atom
-- And implement Eq

type ExitType = Tuple3 Atom Pid Atom

handle :: ExitType -> _
handle = uncurry3 \exit pid res ->
  if res == Atom "normal" then ... else ...
```

So my intention is to publish a package which makes such an `Atom` type, and implement an optimisation in the compiler for atom literals

**Comment by lostbean** [comment]
My initial thought was similar to `type ExitType = Tuple3 Atom Pid Atom` but it seemed to open. This `Atom` representation introduces some _Atom blidness_, to avoid that we could think on something like `data Atom (s :: Symbol) = Atom` and try to get some compile time reflection to generate the erlang atom. The challenge would remain how to represent a sum type of atoms (e.g. `Atom ("normal" | "kill")`).

I'm trying something more idiomatic by adding a lightweight runtime conversion (see 4e02f9229693d0515ac86c04edfa05b51be89adc). @nwolverson what do you think?

**Comment by nwolverson** [comment]
I'll take a detailed look later, but to be clear my suggested type was for an internal representation for typing the FFI definition, which can be exposed with a nicer type to be exported from the library :)

Definitely some symbol stuff would be good.

**Comment by lostbean** [comment]
Thanks for clarifying. So, I think we are basically proposing similar solutions here :) The only difference is where to place the coupling between high and low level representations (directly on FFI or on the PS side), right? IMO, doing it directly on the FFI results in cleaner erlang code with fewer indirections.

**Comment by nwolverson** [comment]
Certainly have no problem with the way you do it in the linked commit. Please do submit a PR when you're happy with what you have

**Comment by nwolverson** [comment]
(Finally) implemented trapped exits in #6.
